cpp/unreal_plugins/UrdfBot/Source/UrdfParser.h

```diff
@@ -0,0 +1,142 @@
+//
+// Copyright(c) 2022 Intel. Licensed under the MIT License <http://opensource.org/licenses/MIT>.
+//
+
+#pragma once
+
+#include <Eigen/Dense>
+#include <map>
+#include <vector>
+
+#include <CoreMinimal.h>
+
+class FXmlNode;
+
+struct UrdfJointDesc;
+struct UrdfLinkDesc;
+struct UrdfMaterialDesc;
+
+enum UrdfGeometryType
+{
+    INVALID_GEOMETRY = 0,
+    BOX,
+    SPHERE,
+    CYLINDER,
+    MESH
+};
+
+enum UrdfJointType
+{
+    // Some of these (such as FIXED) are already macros
+    INVALID_TYPE = 0,
+    REVOLUTE_TYPE,
+    CONTINUOUS_TYPE,
+    PRISMATIC_TYPE,
+    FIXED_TYPE,
+    FLOATING_TYPE,
+    PLANAR_TYPE
+};
+
+struct UrdfGeometry
+{
+    UrdfGeometryType type_;
+    FVector size_; // requried by BOX
+    float radius_; // requried by SPHERE, CYLINDER
+    float length_; // requried by CYLINDER
+    std::string file_name_; // requried by MESH
+};
+
+struct UrdfLinkDesc
+{
+    std::string name_;
+
+    // non-derived data (inertial, visual, etc)...
+    FTransform origin_ = FTransform::Identity;
+    float mass_;
+    Eigen::Matrix3f inertia_;
+    FTransform visual_origin_ = FTransform::Identity;
+    UrdfGeometry visual_geometry_;
+    FTransform collision_origin_ = FTransform::Identity;
+    UrdfGeometry collision_geometry_;
+    std::string material_name_;
+
+    // derived data
+    UrdfLinkDesc* parent_link_desc_ = nullptr;
+    UrdfJointDesc* parent_joint_desc_ = nullptr;
+
+    std::vector<UrdfLinkDesc*> child_link_descs_ = {};
+    std::vector<UrdfJointDesc*> child_joint_descs_ = {};
+
+    UrdfMaterialDesc* material_desc_ = nullptr;
+};
+
+struct UrdfJointDesc
+{
+    std::string name_;
+
+    // non-derived data (origin, axis, etc)...
+    UrdfJointType type_;
+    FTransform origin_ = FTransform::Identity;
+    std::string parent_link_name_;
+    std::string child_link_name_;
+    FVector axis_ = FVector(1, 0, 0);
+    // dynamics
+    float damping_ = 0;
+    float friction_ = 0;
+    //limits
+    float lower_ = 0;
+    float upper_ = 0;
+    float effort_ = 0;
+    float velocity_ = 0;
+
+    // derived data
+    UrdfLinkDesc* parent_link_desc_ = nullptr;
+    UrdfLinkDesc* child_link_desc_ = nullptr;
+};
+
+// material node to describe unreal material that might shared by multiple links
+struct UrdfMaterialDesc
+{
+    std::string name_;
+
+    // non-derived data
+    FVector4 color_;
+    std::string file_name_;
```

Comment:
- call the filename parameter `texture_` to match the existing convention in the rest of these structs

Reply:
fixed as suggested.

We might need to think again about the material node design. 

For links using mesh geometry, some of the texture info are stored in mesh.dae file and the material color is not used. In Unreal, they have a corresponding mesh.uasset, with default material assigned, and we also do not use the material defined in .xml. 

For LoCoBot, there are many meshes do not have texture defined in .dae file, while they still have a default Unreal material. It is because Unreal material only using rgba value only looks wrong and we need to tune the materialmetallic and roughness value to make them look metallic. 

For links using primitive type, they do not have a default material. Therefore we need parse the value from .xml to assign a predefined unreal material. We typically assign frictionless material for those caster links with spherical geometry.

The reason why there is a material node in the top level is that we can share the material instance if multiple geometry have the same material value. It only makes sense if we are creating new material instance dynamically and we are not doing it. Right now we are only loading existing material and Unreal handles the material instance lifetime. I think it is better to remove the top level material node. 

Potentially in the future, we could have a UrdfBotBuilder which parse the raw .urdf file, convert mesh file (.dae, .stl, .obj, etc), and materials (color, texture) to uassets, and generate a corresponding .xml file, or even a Robot.uasset (which might actually be required for SkeletonMesh).

Reply:
I'm not suggesting that we support all possible URDF files when creating our Unreal representation. We might only be able to load a specific subset of URDF files into our Unreal representation (i.e., URDF files with a specific subset of joints, where all mesh information is already in an Unreal format, etc).

I just want the parser to be general enough to represent all possible URDF files, because it isn't that much additional effort or code, and then it (the parser) will be a more useful general-purpose component that we will never need to think about again, even as we expand what types of UDRF files we can support.